### PR TITLE
Add a datepart expression for dfr to be used with dfr with-column

### PR DESCRIPTION
**Pull request by fnordpig**
# Description

Today the only way to extract date parts from a dfr series is the dfr get-* set of commands.  These create a new dataframe with just the datepart in it, which is almost entirely useless.  As far as I can tell there's no way to append it as a series in the original dataframe.  In discussion with fdncred on Discord we decided the best route was to add an expression for modifying columns created in dfr with-column.  These are the way you manipulate series within a data frame.  

I'd like feedback on this approach - I think it's a fair way to handle things.  An example to test it would be:

```[[ record_time]; [ (date now)]]  | dfr into-df | dfr with-column [ ((dfr col record_time) | dfr datepart nanosecond | dfr as "ns" ), (dfr col record_time | dfr datepart second | dfr as "s"), (dfr col record_time | dfr datepart minute | dfr as "m"), (dfr col record_time | dfr datepart hour | dfr as "h") ]```

I'm also proposing we deprecate the dfr get-* commands.  I've not been able to figure out any meaningful way they could ever be useful, and this approach makes more sense by attaching the extracted date part to the row in the original dataframe as a new column.

# User-Facing Changes

add in dfr datepart as an expression

# Tests + Formatting
Need to add some better assertive tests.  I'm also not sure how to properly write the test_dataframe at the bottom, but will revisit as part of this PR.  Wanted to get feedback early.

# After Submitting

### CODE REVIEW

**Comment by fnordpig** [comment]
Sorry I seem to have a dangling commit showing up here from a prior PR

**Comment by fdncred** [comment]
Thanks for this PR!

This isn't meant to take away from this PR, just to mention what was said in discord, this may be another way to achieve the same result.

The new df has the same order of rows as initial df by pola.rs design. You can, actually, insert it back using 'dfr with-column', though you will need to use 'dfr into-df' here and there. 
```
❯ let $t1_df1 = ([[ts const]; [(date now) 1] [((date now) + 50.5sec) 2]]); $t1_df1
┏━━━━━━━━━━━━━┳━━━━━━━┓
┃     ts      ┃ const ┃
┣━━━━━━━━━━━━━╋━━━━━━━┫
┃ now         ┃     1 ┃
┃ in a minute ┃     2 ┃
┗━━━━━━━━━━━━━┻━━━━━━━┛

❯ $t1_df1 | dfr into-df | dfr with-column ($t1_df1 | dfr into-df | dfr select ts | dfr get-nanosecond | dfr rename ts ts_n)
┏━━━┳━━━━━━━━━━━━━━━┳━━━━━━━┳━━━━━━━━━━━┓
┃ # ┃      ts       ┃ const ┃   ts_n    ┃
┣━━━╋━━━━━━━━━━━━━━━╋━━━━━━━╋━━━━━━━━━━━┫
┃ 0 ┃ now           ┃     1 ┃ 434000000 ┃
┃ 1 ┃ in 45 seconds ┃     2 ┃ 934000000 ┃
┗━━━┻━━━━━━━━━━━━━━━┻━━━━━━━┻━━━━━━━━━━━┛
```

File: crates/nu-cmd-dataframe/src/dataframe/expressions/datepart.rs [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+            "second" => expr_dt.second(),
+            "millisecond" => expr_dt.millisecond(),
+            "microsecond" => expr_dt.microsecond(),
+            "nanosecond" => expr_dt.nanosecond(),
+            _ => {
+                return Err(ShellError::UnsupportedInput(
+                    format!("{} is not a valid datepart, expected one of year, month, day, hour, minute, second, millisecond, microsecond, nanosecond", part),
+                    "value originates from here".to_string(),
+                    call.head,
+                    call.head, // need to do better here
```
Comment (fdncred) [inline crates/nu-cmd-dataframe/src/dataframe/expressions/datepart.rs:96]: need to do better? i guess that means you're not using unsupportedinput right?

**Review by fdncred — COMMENTED** [review]
I think this looks pretty good. I'd like to fix whatever the CI is complaining about and address the issues I've mentioned, but after that, I'll be ready to land this. Good work!

I want to hold off deprecating the get-* for a while though. Thanks!

I tested your one-liner and it works great!

File: crates/nu-cmd-dataframe/src/dataframe/expressions/datepart.rs (hunk 1 above)
Comment (fnordpig) [reply crates/nu-cmd-dataframe/src/dataframe/expressions/datepart.rs:96]: Sorry, here I think the span isn't right - I think it should be the span for the part of the datepart, but I wasn't super clear on how to traverse to it and wanted early directional feedback.

File: crates/nu-cmd-dataframe/src/dataframe/expressions/datepart.rs [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+                    span: Span::test_data(),
+                };
+
+                Some(record)
+            },
+        }]
+    }
+
+    fn search_terms(&self) -> Vec<&str> {
+        vec!["datepart", "date", "year", "month", "week", "weekday", "quarter", "day", "hour", "minute", "second", "millisecond", "microsecond", "nanosecond"]
```
[the remark this answers is not in the document]
Comment (fnordpig) [reply crates/nu-cmd-dataframe/src/dataframe/expressions/datepart.rs:65]: Will do.

File: crates/nu-cmd-dataframe/src/dataframe/expressions/datepart.rs (hunk 1 above)
Comment (fdncred) [reply crates/nu-cmd-dataframe/src/dataframe/expressions/datepart.rs:110]: I'd have to play a bit but I think if you get the `part` as a `Spanned<String>`, you'd have the span for the part.
```rust
let part: Spanned<String> = call.req(engine_state, stack, 0)?;
```
There's examples of how to get a spanned string in other commands i'm sure.

**Comment by fdncred** [comment]
>  It's tricky to figure out in what context what will work

Agreed. It would be nice to have that in the help somehow. I think it's in the category, but it should be easier to figure out.
image

**Comment by fnordpig** [comment]
So, I'm trying to write tests and there's a bit of a challenge.  I need to pull in IntoDateTime, but due to the cratificiation it looks like all the references to IntoDateTime in dataframe are configured away:
c55b5c0a55896f998d01a0bed6d53fa8a90381c2

I was trying to use ``date now``, but it similarly depends on adding a new dependency to the dataframe crate (nu-commands)

**Comment by fdncred** [comment]
You're right. I forgot about the cratification and how it disabled datetime stuff now. We can save that for another PR. I wonder if you can do some tests like where you hard code some datetime and then just get-* off of that hard coded date. Is that possible?

**Comment by fnordpig** [comment]
OK found a way to get it to work, the ``dfr as_datetime`` command is available to convert a string into a datetime without touching any external crates.  I added one example test, but I'll add a few more shortly.

**Comment by fdncred** [comment]
Making draft until you're ready.

**Comment by fnordpig** [comment]
This turned out to be more complex than I thought, in several places there was an assumption that everything in a dfr is milliseconds despite supporting nanoseconds in both the conversion routines and in chrono::DateTime<>.  This adds some tests for nanoseconds where it was most obviously not working as well as adjusts things to ensure they work consistently.  I didn't move the sql types to nanoseconds because my research says the maximum precision expected there is microseconds.

**Comment by fdncred** [comment]
Thanks for all the hard work here! It's nice to have more dataframe contributors!!! ❤️

**Pull request merged**